File: frontend/src/api/allocations.ts
// Work Order Allocation API (Phase 10 — PROJECT_PROFILE §8.12, §18;
// GUI_DESIGN §10).
//
// The receiving confirmation that follows a `STOCKED` arrival at the
// Stockroom station: the server's canonical allocation suggestion (a
// read), and the confirmed allocation (one command, idempotent per
// `device_event_id`). Allocation is a record of its own — it never
// references a Movement or a Quantity Flow and never changes the
// Movement history.
//
// Production-safe: no mock data, no framework imports.

import { apiRequest, apiRequestWithStatus } from './client';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** One outstanding demand line in the canonical demand ordering. */
export interface SuggestedAllocationLine {
  workOrderId: number;
  workOrderNumber: string | null;
  receivedDate: string;
  workOrderDemandId: number;
  priorityRank: number | null;
  dueDate: string | null;
  requestedQuantity: number;
  previouslyAllocatedQuantity: number;
  remainingShortage: number;
  /** The server's proposal for THIS line (0 when the quantity does not
   * reach it) — the operator may adjust it before confirming. */
  proposedQuantity: number;
}

export interface AllocationSuggestion {
  partNumber: string;
  /** The quantity the suggestion was computed for — capped by the
   * server at the available stocked quantity. */
  quantity: number;
  stockedQuantity: number;
  activeAllocatedQuantity: number;
  availableStockedQuantity: number;
  proposedTotal: number;
  /** Quantity no outstanding demand can take — it stays in stock. */
  unallocatedQuantity: number;
  /** Every outstanding line, canonical order, proposed-0 lines included. */
  lines: SuggestedAllocationLine[];
}

export interface AllocationLine {
  workOrderDemandId: number;
  quantity: number;
}

export interface AllocationInput {
  partNumber: string;
  /**
   * The explicit quantity being allocated — at the Stockroom the
   * just-stocked quantity the operator confirmed, never the PN's whole
   * available stock. The lines must sum to exactly it (the server
   * refuses anything else with nothing written), and the server judges
   * it against the available stocked quantity under its lock: a figure
   * that shrank since the dialog opened is refused, nothing written.
   */
  allocationQuantity: number;
  lines: AllocationLine[];
  /** The Stockroom station confirming the receiving allocation. */
  stationId: string | null;
  /** Client-generated UUID, reused verbatim on every retry of the SAME
   * confirmed intent (idempotency key). */
  deviceEventId: string;
}

export interface AllocationRow {
  allocationId: number;
  workOrderDemandId: number;
  workOrderId: number;
  quantity: number;
  isManualOverride: boolean;
}

export interface AllocationResult {
  partNumber: string;
  allocationQuantity: number;
  rows: AllocationRow[];
  /** Work Orders this confirmation completed — every demand line fully
   * allocated (server-derived). */
  completedWorkOrderIds: number[];
  deviceEventId: string;
  /** false when the server replayed an already committed allocation. */
  created: boolean;
}

// ---------------------------------------------------------------------------
// Wire mapping
// ---------------------------------------------------------------------------

interface SuggestedLineWire {
  work_order_id: number;
  work_order_number: string | null;
  received_date: string;
  work_order_demand_id: number;
  priority_rank: number | null;
  due_date: string | null;
  requested_quantity: number;
  previously_allocated_quantity: number;
  remaining_shortage: number;
  proposed_quantity: number;
}

interface SuggestionWire {
  part_number: string;
  quantity: number;
  stocked_quantity: number;
  active_allocated_quantity: number;
  available_stocked_quantity: number;
  proposed_total: number;
  unallocated_quantity: number;
  lines: SuggestedLineWire[];
}

interface AllocationRowWire {
  allocation_id: number;
  work_order_demand_id: number;
  work_order_id: number;
  quantity: number;
  is_manual_override: boolean;
}

interface AllocationResultWire {
  part_number: string;
  allocation_quantity: number;
  rows: AllocationRowWire[];
  completed_work_order_ids: number[];
  device_event_id: string;
}

// ---------------------------------------------------------------------------
// Calls
// ---------------------------------------------------------------------------

/**
 * The canonical allocation suggestion for a PN — a read, nothing is
 * recorded. `quantity` is the quantity about to be allocated (the
 * just-stocked quantity); omitted, the server suggests for the whole
 * available stock.
 */
export async function getAllocationSuggestion(
  partNumber: string,
  quantity?: number,
): Promise<AllocationSuggestion> {
  const params = new URLSearchParams({ part_number: partNumber });
  if (quantity !== undefined) params.set('quantity', String(quantity));
  const wire = await apiRequest<SuggestionWire>(
    `/api/allocations/suggestion?${params.toString()}`,
  );
  return {
    partNumber: wire.part_number,
    quantity: wire.quantity,
    stockedQuantity: wire.stocked_quantity,
    activeAllocatedQuantity: wire.active_allocated_quantity,
    availableStockedQuantity: wire.available_stocked_quantity,
    proposedTotal: wire.proposed_total,
    unallocatedQuantity: wire.unallocated_quantity,
    lines: wire.lines.map((line) => ({
      workOrderId: line.work_order_id,
      workOrderNumber: line.work_order_number,
      receivedDate: line.received_date,
      workOrderDemandId: line.work_order_demand_id,
      priorityRank: line.priority_rank,
      dueDate: line.due_date,
      requestedQuantity: line.requested_quantity,
      previouslyAllocatedQuantity: line.previously_allocated_quantity,
      remainingShortage: line.remaining_shortage,
      proposedQuantity: line.proposed_quantity,
    })),
  };
}

/**
 * Record the confirmed allocation. Resolves ONLY when the server
 * confirmed the write: 201 fresh, 200 on an idempotent replay of the
 * same `deviceEventId` + same intent. Every refusal — lines not adding
 * up to the allocation quantity, a line beyond its remaining shortage,
 * an allocation quantity the available stock no longer covers — is an
 * `ApiError` and nothing was recorded.
 */
export async function confirmAllocation(
  input: AllocationInput,
): Promise<AllocationResult> {
  const { status, data } = await apiRequestWithStatus<AllocationResultWire>(
    '/api/allocations',
    {
      method: 'POST',
      body: {
        part_number: input.partNumber,
        allocation_quantity: input.allocationQuantity,
        lines: input.lines.map((line) => ({
          work_order_demand_id: line.workOrderDemandId,
          quantity: line.quantity,
        })),
        ...(input.stationId !== null ? { station_id: input.stationId } : {}),
        device_event_id: input.deviceEventId,
      },
    },
  );
  return {
    partNumber: data.part_number,
    allocationQuantity: data.allocation_quantity,
    rows: data.rows.map((row) => ({
      allocationId: row.allocation_id,
      workOrderDemandId: row.work_order_demand_id,
      workOrderId: row.work_order_id,
      quantity: row.quantity,
      isManualOverride: row.is_manual_override,
    })),
    completedWorkOrderIds: data.completed_work_order_ids,
    deviceEventId: data.device_event_id,
    created: status === 201,
  };
}

File: backend/app/api/allocations.py
"""Work Order Allocation endpoints (Phase 10 — PROJECT_PROFILE §8.12, §18; GUI_DESIGN §10).

Thin routes over `app.application.allocations`: request schemas
validate shape only (``extra="forbid"``), every rule and transaction
lives in the Application layer, and the central handlers translate the
typed errors.

Surface:

- ``GET  /allocations/suggestion?part_number=…&quantity=…`` — the
  canonical allocation suggestion for a PN (a read): the PN's stocked,
  allocated and available stocked quantity, and every outstanding
  demand line in the canonical demand ordering with the requested,
  previously allocated, remaining shortage and proposed quantity. The
  quantity defaults to the whole available stocked quantity and is
  capped at it. The Stockroom station calls it with the just-stocked
  quantity right after the ``STOCKED`` write.
- ``POST /allocations`` — the confirmed allocation of one PN's stocked
  quantity to demand lines (the receiving confirmation with
  ``station_id`` set — a Stockroom station; a Management allocation
  without it). 201 fresh, 200 on an idempotent replay of the same
  ``device_event_id`` + same intent, 422 when the lines do not sum to
  the explicit ``allocation_quantity`` (or when it is missing), 409 on
  a mismatched reuse, on a line beyond its remaining shortage, or on an
  allocation quantity the available stocked quantity no longer covers
  (a stale figure) — every refusal writes nothing.
- ``POST /allocations/{allocation_id}/reversals`` — the auditable
  adjustment: takes one allocation back with a mandatory reason (a
  smaller allocation is a reversal plus a new confirmation). 201 /
  200 / 409 as above; 409 when already reversed (also under a race).
- ``GET  /allocations?part_number=…&work_order_demand_id=…&work_order_id=…``
  — the allocation rows (allocations and reversals) for audit display.

No authorization is enforced or simulated (Phase 14): the ``actor``
field is never client-writable — audit rows stay NULL until an
authenticated identity exists.
"""

import datetime
from typing import Literal

from fastapi import APIRouter, Response
from pydantic import BaseModel, ConfigDict, StrictInt

from app.api.dependencies import SessionDep
from app.application import allocations

router = APIRouter(prefix="/api")


class SuggestedLineResponse(BaseModel):
    work_order_id: int
    work_order_number: str | None
    received_date: datetime.date
    work_order_demand_id: int
    priority_rank: int | None
    due_date: datetime.date | None
    requested_quantity: int
    previously_allocated_quantity: int
    remaining_shortage: int
    proposed_quantity: int


class AllocationSuggestionResponse(BaseModel):
    part_number: str
    quantity: int
    stocked_quantity: int
    active_allocated_quantity: int
    available_stocked_quantity: int
    proposed_total: int
    # Quantity no outstanding demand can take: it stays in stock.
    unallocated_quantity: int
    lines: list[SuggestedLineResponse]


@router.get("/allocations/suggestion")
def get_allocation_suggestion(
    session: SessionDep, part_number: str, quantity: int | None = None
) -> AllocationSuggestionResponse:
    suggestion = allocations.suggest_allocation(session, part_number=part_number, quantity=quantity)
    return AllocationSuggestionResponse(
        part_number=suggestion.part_number,
        quantity=suggestion.quantity,
        stocked_quantity=suggestion.stocked_quantity,
        active_allocated_quantity=suggestion.active_allocated_quantity,
        available_stocked_quantity=suggestion.available_stocked_quantity,
        proposed_total=suggestion.proposed_total,
        unallocated_quantity=suggestion.unallocated_quantity,
        lines=[
            SuggestedLineResponse(
                work_order_id=line.work_order.id,
                work_order_number=line.work_order.work_order_number,
                received_date=line.work_order.received_date,
                work_order_demand_id=line.demand.id,
                priority_rank=line.demand.priority_rank,
                due_date=line.demand.due_date,
                requested_quantity=line.requested_quantity,
                previously_allocated_quantity=line.previously_allocated_quantity,
                remaining_shortage=line.remaining_shortage,
                proposed_quantity=line.proposed_quantity,
            )
            for line in suggestion.lines
        ],
    )


class AllocationLineRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    work_order_demand_id: int
    quantity: StrictInt


class AllocationRequest(BaseModel):
    """The confirmed allocation (the receiving confirmation, or Management)."""

    model_config = ConfigDict(extra="forbid")

    part_number: str
    # The explicit quantity being allocated — at the Stockroom the
    # just-stocked quantity the operator confirmed. The lines must sum
    # to exactly it; the available stocked quantity must still cover
    # it when the command is judged (a stale figure is refused).
    allocation_quantity: StrictInt
    lines: list[AllocationLineRequest]
    # The Stockroom station confirming the receiving allocation; omitted
    # for a Management allocation.
    station_id: str | None = None
    reason: str | None = None
    device_event_id: str


class AllocationRowResponse(BaseModel):
    allocation_id: int
    work_order_demand_id: int
    work_order_id: int
    part_number: str
    quantity: int
    source: Literal["STOCKROOM", "MANAGEMENT"]
    is_manual_override: bool
    allocation_reason: str | None
    # Set on a reversal row: the allocation it takes back.
    reverses_allocation_id: int | None
    station_id: str | None
    actor_reference: str | None
    allocated_at: datetime.datetime
    command_sequence: int


class AllocationResponse(BaseModel):
    kind: Literal["ALLOCATE", "REVERSE_ALLOCATION"]
    part_number: str
    # The quantity allocated (or, on a reversal, taken back).
    allocation_quantity: int
    rows: list[AllocationRowResponse]
    # Work Orders this command completed (every line fully allocated)
    # or reopened (a reversal) — the derived completion effect.
    completed_work_order_ids: list[int]
    reopened_work_order_ids: list[int]
    device_event_id: str


def _row(row: allocations.AllocationRow) -> AllocationRowResponse:
    return AllocationRowResponse(
        allocation_id=row.allocation_id,
        work_order_demand_id=row.work_order_demand_id,
        work_order_id=row.work_order_id,
        part_number=row.part_number,
        quantity=row.quantity,
        source="STOCKROOM" if row.source == "STOCKROOM" else "MANAGEMENT",
        is_manual_override=row.is_manual_override,
        allocation_reason=row.allocation_reason,
        reverses_allocation_id=row.reverses_allocation_id,
        station_id=row.station_id,
        actor_reference=row.actor_reference,
        allocated_at=row.allocated_at,
        command_sequence=row.command_sequence,
    )


def _response(result: allocations.AllocationResult) -> AllocationResponse:
    return AllocationResponse(
        kind="ALLOCATE" if result.kind == "ALLOCATE" else "REVERSE_ALLOCATION",
        part_number=result.part_number,
        allocation_quantity=result.allocation_quantity,
        rows=[_row(row) for row in result.rows],
        completed_work_order_ids=result.completed_work_order_ids,
        reopened_work_order_ids=result.reopened_work_order_ids,
        device_event_id=result.device_event_id,
    )


@router.post("/allocations")
def confirm_allocation(
    body: AllocationRequest, session: SessionDep, response: Response
) -> AllocationResponse:
    result = allocations.confirm_allocation(
        session,
        part_number=body.part_number,
        allocation_quantity=body.allocation_quantity,
        lines=[line.model_dump() for line in body.lines],
        station_id=body.station_id,
        reason=body.reason,
        device_event_id=body.device_event_id,
    )
    response.status_code = 201 if result.created else 200
    return _response(result)


class AllocationReversalRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reason: str
    station_id: str | None = None
    device_event_id: str


@router.post("/allocations/{allocation_id}/reversals")
def reverse_allocation(
    allocation_id: int,
    body: AllocationReversalRequest,
    session: SessionDep,
    response: Response,
) -> AllocationResponse:
    result = allocations.reverse_allocation(
        session,
        allocation_id=allocation_id,
        reason=body.reason,
        station_id=body.station_id,
        device_event_id=body.device_event_id,
    )
    response.status_code = 201 if result.created else 200
    return _response(result)


class AllocationRecordResponse(BaseModel):
    id: int
    part_number: str
    work_order_demand_id: int
    quantity: int
    source: str
    is_manual_override: bool
    allocation_reason: str | None
    reverses_allocation_id: int | None
    station_id: str | None
    actor_reference: str | None
    allocated_at: datetime.datetime
    device_event_id: str
    command_sequence: int


@router.get("/allocations")
def list_allocations(
    session: SessionDep,
    part_number: str | None = None,
    work_order_demand_id: int | None = None,
    work_order_id: int | None = None,
) -> list[AllocationRecordResponse]:
    return [
        AllocationRecordResponse(
            id=row.id,
            part_number=row.part_number,
            work_order_demand_id=row.work_order_demand_id,
            quantity=row.quantity,
            source=row.source,
            is_manual_override=row.is_manual_override,
            allocation_reason=row.allocation_reason,
            reverses_allocation_id=row.reverses_allocation_id,
            station_id=row.station_id,
            actor_reference=row.actor_reference,
            allocated_at=row.allocated_at,
            device_event_id=row.device_event_id,
            command_sequence=row.command_sequence,
        )
        for row in allocations.list_allocations(
            session,
            part_number=part_number,
            work_order_demand_id=work_order_demand_id,
            work_order_id=work_order_id,
        )
    ]
